man/fileLogger.3
.\" Written  Oct 2012 by Olivier Charloton (oliviers.open.source.softwares@gmail.com)
.\" heavily based on the work of Steve Greenland (stevegr@neosoft.com) on syslog man.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.TH FILELOGGER 3 2012-10-06 "dbgflags" "Dbgflags Programmer's Manual"
.SH NAME
openLogFile, closeLogFile, fileLogger, vfileLogger, setFileLoggerMaxSize, setFileLoggerMaxDuration, setFileLoggerDirectory  \- log messages to files.
.SH SYNOPSIS
.B #include <dbgflags/loggers.h>
.sp
.BI "void openLogFile(const char *" ident ", int " option ", int " facility ");"
.br
.BI "void fileLogger(int " priority ", const char *" format ", ...);"
.br
.BI "int setFileLoggerMaxSize(const size_t " maxSizeInBytes ");"
.br
.BI "int setFileLoggerMaxDuration(const time_t " maxDurationInSeconds ");"
.br
.BI "int setFileLoggerDirectory(const char *" directory ");"
.br
.BI "void closeLogFile(void);"
.sp
.B #include <stdarg.h>
.sp
.BI "void vfileLogger(int " priority ", const char *" format ", va_list " ap ");"
.br
.SH DESCRIPTION
.BR openLogFile ()
create a new log file for a program.  The string pointed
to by
.I ident
is prepended to every message, and is typically set to the program name.
The
.I option
argument specifies flags which control the operation of
.BR openLogFile ()
and subsequent calls to
.BR fileLogger ().
The
.I facility
argument establishes a default to be used if
none is specified in subsequent calls to
.BR fileLogger ().
Values for
.I option
and
.I facility
are given below.  The use of 
.BR openLogFile ()
is optional; it will automatically be called by
.BR fileLogger ()
if necessary, in which case 
.I ident
will default to NULL.
.sp
.BR fileLogger ()
generates a log message, which will be written to a file.
The
.I priority
argument is formed by ORing the
.I facility
and the 
.I level
values (explained below).  The remaining arguments
are a 
.IR format ,
as in 
.BR printf (3)
and any arguments required by the 
.IR format , 
except that the two character sequence %m will be replaced by
the error message string 
.IR strerror ( errno ).
A trailing newline is added when needed.

The function
.BR vfileLogger ()
performs the same task as
.BR fileLogger ()
with the difference that it takes a set of arguments which have
been obtained using the
.BR stdarg (3)
variable argument list macros.

.BR setFileLoggerMaxSize()
set the maximum size in bytes of a log file. When this size is reached a new log file will be created. 
Remark: messages are not splitted between log files so the log file's size will be greater or equal to the maximum size.

.BR setFileLoggerMaxDuration()
set the maximum duration in seconds of a log file. When this duration is reached a new log file will be created. 
Remark: messages are not splitted between log files so the log file's duration will be greater or equal to the maximum duration.

.BR setFileLoggerDirectory()
set the directory where the log files will be located.

.BR closeLogFile ()
closes the descriptor file being used to write to the log file.  The use of
.BR closeLogFile ()
is optional.

.SH PARAMETERS
This section lists the parameters used to set the values of 
.IR option , " facility" ", and " priority 
(sames as 
.BR syslog(3)
).
.SS option
The
.I option
argument to 
.BR openLogFile ()
is an OR of any of these:
.TP
.B LOG_CONS
Write directly to system console if there is an error while logging to
a file.
.TP
.B LOG_NDELAY
Create the log file immediately (normally, the file is created when
the first message is logged).
.TP
.B LOG_NOWAIT
Ignored.
.TP
.B LOG_ODELAY
The converse of
.BR LOG_NDELAY ;
Ignored too.
.TP
.B LOG_PERROR
(Not in POSIX.1-2001.) Print to stderr as well.
.TP
.B LOG_PID
Include PID with each message.
.TP
.B LOG_TID
(dbgflags's extension) Include Thread's identity (TID) with each message.
.TP
.B LOG_RDTSC
(dbgflags's extension) Log the number of clock cycles, since the CPU was powered up or reset, with each message. 
Warning: this value is different between every core and CPU, when running on more then one, use the LOG_CLOCK instead.
.TP 
.B LOG_LEVEL
(dbgflags's extension) Log the message level as a string with each message.
.TP
.B LOG_CLOCK
(dbgflags's extension) Log a time, that represents monotonic time since some unspecified starting point, with each message.
.TP 
.B LOG_FILE_ROTATE
(dbgflags's extension to manage log files) Use two files to log messages, when the threshold (set by 
.BR setFileLoggerMaxSize(3) 
or 
.BR setFileLoggerMaxDuration(3)) 
is reached, 
the current log file is closed then renamed to old and a new one is created (the previous old file if any is deleted). 
.TP 
.B LOG_FILE_HISTO
(dbgflags's extension to manage log files) Each log file's name will contains the current start time (as  the  number of seconds since the Epoch, 1970-01-01 00:00:00 +0000 (UTC)) and previous files will not deleted when the defined threshold will be reached.
.TP 
.B LOG_FILE_DURATION
(dbgflags's extension to manage log files) Enable the log file swap on duration (default is on size).
.TP 
.B LOG_FILE_WITHOUT_SYNC
(dbgflags's extension to manage log files) Disabled the O_SYNC flag (synchronous I/O to avoid data lost in case of program crash) of the file creation parameters, see open(2) for details. Remark: this option has been added on request for performances reasons, but be aware that the last message written in the log file may  be no more the last one when set.
.TP 
.B LOG_FILE_SYNC_ON_ERRORS_ONLY
(dbgflags's extension to manage log files) Disabled the O_SYNC flag (synchronous I/O to avoid data lost in case of program crash) of the file creation parameters, see open(2) for details, but a data flush of the log file will be performed if the importance of the message is upper or equal to error. Remark: this flag offers a compromise with the previous one.

.SS facility 
The 
.I facility
argument is used to specify what type of program is logging the message.
This lets the configuration file specify that messages from different
facilities will be handled differently.
.TP
.B LOG_AUTH
security/authorization messages (DEPRECATED Use 
.B LOG_AUTHPRIV 
instead)
.TP
.B LOG_AUTHPRIV
security/authorization messages (private) 
.TP
.B LOG_CRON
clock daemon 
.RB ( cron " and " at )
.TP
.B LOG_DAEMON
system daemons without separate facility value
.TP
.B LOG_FTP
ftp daemon
.TP
.B LOG_KERN
kernel messages
.TP
.BR LOG_LOCAL0 " through " LOG_LOCAL7
reserved for local use
.TP
.B LOG_LPR
line printer subsystem
.TP
.B LOG_MAIL
mail subsystem
.TP
.B LOG_NEWS
USENET news subsystem
.TP
.B LOG_SYSLOG
messages generated internally by
.B syslogd 
.TP
.BR LOG_USER " (default)"
generic user-level messages 
.TP
.B LOG_UUCP
UUCP subsystem

.SS level
This determines the importance of the message.  The levels are, in order
of decreasing importance:
.TP
.B LOG_EMERG
system is unusable
.TP
.B LOG_ALERT
action must be taken immediately
.TP
.B LOG_CRIT
critical conditions
.TP
.B LOG_ERR
error conditions
.TP
.B LOG_WARNING
warning conditions
.TP
.B LOG_NOTICE
normal, but significant, condition
.TP
.B LOG_INFO
informational message
.TP
.B LOG_DEBUG
debug-level message
.LP
The functions
.BR setlogmaskex (3)
or
.BR setlogmask (3)
could be used to restrict logging to specified levels only.
.SH NOTES
The parameter
.I ident
in the call of
.BR openLogFile ()
is stored as-is. Thus, if the string it points to
is changed,
.BR fileLogger ()
may start prepending the changed string, and if the string
it points to ceases to exist, the results are undefined.
Use a string constant.
.LP
Never pass a string with user-supplied data as a format, use the folloing instead:
.RS
fileLogger(priority, "%s", string);

.SH "SEE ALSO"
.BR logger (1),
.BR syslog (3),
.BR syslogex (3),
.BR setlogmask (3),



NEWS
- A few corrections to build on a Debian 6 64bits.
- pkgconfig test program build corrected when the library hase not been already installed.
- the gettid internal function use the syscall function (instead of the "int 0x80") when building on a 2.6.x Linux kernel.
- Buffer overflow correction in consoleLogger & fileLogger (use a stream buffer now instead of a buffer on the stack).
  and new test programs to check the library's logger with message of size up to 4096 bytes.
- thread cancellation points management in fileLogger added (useless in the consoleLogger).
- added instructions to free the used buffer in the thread cancellation points management of the syslogex logger (glibc bug ?)
- logger sending trace to one file per thtread

Remarques:
- Files'repository moved to GIT and "public" version published to github
- using redmine to manage evolutions (ER) and anomalies reports (AR).


libdbgflags/fileLogger.c
#include "config.h"
#define _XOPEN_SOURCE 600

#include <stdio.h>
#include <stdarg.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <pthread.h>
#include <time.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <linux/limits.h>
#include <dbgflags/syslogex.h>
#include <dbgflags/dbgflags.h>
#include <dbgflags/loggers.h>
#include "system.h"

#define DEBUG_LOG_HEADER "fileLogger"
#define LOGGER consoleLogger
#define LOG_OPTS  0


#ifdef _DEBUGFLAGS_H_
static DebugFlags debugFlags =
{
		"fileLogger",
		{
				"file"
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""}
		,0xFFFFFFFF
};
#define FLAG_FILE  FLAG_1

#else /* _DEBUGFLAGS_H_ */
#define FILTER
#define FLAG_FILE
#endif /* _DEBUGFLAGS_H_ */

#include <dbgflags/debug_macros.h>

#define MODULE_FLAG FLAG_FILE

static int	LogStat = 0;		/* status bits, set by openlog() */
static const char *LogTag = NULL;		/* string to tag the entry with */
static int	LogFacility = LOG_USER;	/* default facility code */
//static int LogMask = 0xff; /* mask of priorities to be logged */
static size_t maxSize = (1024*100);
static int logFile = -1;
static char fullProcessName[PATH_MAX];
static char fullFileName[PATH_MAX];
static char directory[PATH_MAX];
static char *processName = NULL;
static char *fileDirectory = NULL;
static pthread_mutex_t fileLock = PTHREAD_RECURSIVE_MUTEX_INITIALIZER_NP;
static size_t fileSize = 0;
static time_t maxDuration = 24*60*60; /* 24 hours */
static time_t startTime = 0;

#include "filesUtils.h"
#include "setFullFileName.h"

static inline int createFile(void) {

#ifdef _DEBUGFLAGS_H_
	{
		static unsigned int registered = 0;
		if (unlikely(0 == registered)) {
			registered = 1; /* dirty work around to avoid deadlock: syslogex->register->syslogex */
			registered = (registerLibraryDebugFlags(&debugFlags) == EXIT_SUCCESS);
		}
	}
#endif /*_DEBUGFLAGS_H_*/

	int error = pthread_mutex_lock(&fileLock);

	if (likely(EXIT_SUCCESS == error)) {
		int internalError = EXIT_SUCCESS;
		int flags = O_WRONLY|O_CREAT|O_TRUNC;

		if (likely((!(LogStat & LOG_FILE_WITHOUT_SYNC)) && (!(LogStat & LOG_FILE_SYNC_ON_ERRORS_ONLY)))) {
			flags |= O_SYNC; /* enable synchronous I/O to avoid data lost in case of crash */
		}

		if (logFile != -1) {
			WARNING_MSG("logFile was NOT NULL");
			if (close(logFile) != 0) {
				internalError = errno;
				ERROR_MSG("close %d error %d (%m)", logFile, internalError);
			}
			logFile = -1;
		}

		if (unlikely(NULL == processName)) {
			setProcessName();
		}
		setFullFileName();

		logFile = open(fullFileName,flags,S_IRUSR|S_IWUSR|S_IRGRP);
		if (likely(logFile != -1)) {
			fileSize = 0;
			if (LOG_FILE_DURATION & LogStat) {
				startTime = time(NULL);
			}
		} else {
			error = errno;
			ERROR_MSG("open %s for creation error %d (%m)", fullFileName, error);
		}

		internalError = pthread_mutex_unlock(&fileLock);
		if (internalError != EXIT_SUCCESS) {
			ERROR_MSG("pthread_mutex_lock fileLock error %d (%s)", internalError, strerror(internalError));
			if (EXIT_SUCCESS == error) {
				error = internalError;
			}
		}

	} else {
		ERROR_MSG("pthread_mutex_lock fileLock error %d (%m)", error);
	}

	return error;
}

void openLogFile(const char *ident, int logstat, int logfac) {
	if (ident != NULL) {
		LogTag = ident;
	}
	LogStat = logstat;
	if (logfac != 0 && (logfac &~ LOG_FACMASK) == 0) {
		LogFacility = logfac;
	}

	if (LogStat & LOG_NDELAY) {
		createFile();
	}
}

typedef struct cleanup_args_ {
	void *buffer;
	//char *logBuffer;
	/* mutex is a static global variable so... */
} cleanup_args;

static void cancel_handler(void *ptr) {
	cleanup_args *args = (cleanup_args *)ptr;
	if ((args != NULL)){
		if (args->buffer != NULL) {
			free(args->buffer);
			args->buffer = NULL;
		}
		/*if (args->logBuffer != NULL) {
	free(args->logBuffer);
	args->logBuffer = NULL;
      }*/
	}
	/* Free the lock.  */
	pthread_mutex_unlock(&fileLock);
}

/* strcpy then strcat function using the same cursor */
static inline void strcpyNcat(char *buffer, const char *source1,const char *source2) {
	register char *p = buffer;

	register const char *s = source1;
	while(*s != '\0') {
		*p = *s;
		p++;
		s++;
	}

	s = source2;
	while(*s != '\0') {
		*p = *s;
		p++;
		s++;
	}

	*p = '\0';
}

void vfileLogger(int priority, const char *format, va_list optional_arguments) {
	int n = 0;
	int error = EXIT_SUCCESS;
	const int saved_errno = errno;
	int internalError = EXIT_SUCCESS;
	const int LogMask = setlogmask(0);

	/* Check priority against setlogmask values. */
	if ((LOG_MASK (LOG_PRI (priority)) & LogMask) != 0) {
		//char logMsg[2048];
		//char logFormat[1024];
		//char *cursor = logFormat;
		char *buf = 0;
		size_t bufsize = 1024;
		FILE *f = open_memstream(&buf, &bufsize);
		if (f != NULL) {
			struct tm now_tm;
			time_t now;

			(void) time(&now);
			/*cursor += strftime(cursor,sizeof(logFormat),"%h %e %T ",localtime_r(&now, &now_tm));*/
			n = strftime(f->_IO_write_ptr,f->_IO_write_end - f->_IO_write_ptr,"%h %e %T ",localtime_r(&now, &now_tm));
			//f->_IO_write_ptr += strftime(f->_IO_write_ptr,f->_IO_write_end - f->_IO_write_ptr,"%h %e %T ",localtime_r(&now, &now_tm));
			f->_IO_write_ptr += n;

			if (LogTag) {
				/*cursor += sprintf (cursor,"%s: ",LogTag);*/
				n += fprintf(f,"%s: ",LogTag);
			}

			if (LogStat & LOG_PID) {
				if (LogStat & LOG_TID) {
					const pid_t tid = gettid();
					/*cursor += sprintf (cursor, "[%d:%d]", (int) getpid (),(int) tid);*/
					n += fprintf(f,"[%d:%d]", (int) getpid (),(int) tid);
				} else {
					/*cursor += sprintf (cursor, "[%d]", (int) getpid ());*/
					n += fprintf(f,"[%d]", (int) getpid ());
				}
			}

			if (LogStat & LOG_RDTSC) {
				const unsigned long long int  t = rdtsc();
				/*cursor += sprintf (cursor, "(%llu)",t);*/
				n += fprintf(f,"(%llu)",t);
			} /* (LogStat & LOG_RDTSC) */

			if (LogStat & LOG_CLOCK) {
#if HAVE_CLOCK_GETTIME
				struct timespec timeStamp;
				if (clock_gettime(CLOCK_MONOTONIC,&timeStamp) == 0) {
					/*cursor += sprintf (cursor, "(%lu.%.9d)",timeStamp.tv_sec,timeStamp.tv_nsec);*/
					n += fprintf(f,"(%lu.%.9d)",timeStamp.tv_sec,timeStamp.tv_nsec);
				} else {
					const int error = errno;
					ERROR_MSG("clock_gettime CLOCK_MONOTONIC error %d (%m)",error);
				}
#else
				static unsigned int alreadyPrinted = 0; /* to avoid to print this error msg on each call */
				if (unlikely(0 == alreadyPrinted)) {
					ERROR_MSG("clock_gettime  not available on this system");
					alreadyPrinted = 1;
				}
#endif
			} /* (LogStat & LOG_CLOCK) */

			if (LogStat & LOG_LEVEL) {
				switch(LOG_PRI(priority)) {
				case LOG_EMERG:
					/*cursor += snprintf(cursor,sizeof(logFormat) - (cursor - logFormat),"* Emergency * %s",format);*/
					n += fprintf(f, "[EMERG]");
					break;
				case LOG_ALERT:
					/*cursor += snprintf(cursor,sizeof(logFormat) - (cursor - logFormat),"* Alert * %s",format);*/
					n += fprintf(f, "[ALERT]");
					break;
				case LOG_CRIT:
					/*cursor += snprintf(cursor,sizeof(logFormat) - (cursor - logFormat),"* Critical * %s",format);*/
					n += fprintf(f, "[CRIT]");
					break;
				case LOG_ERR:
					/*cursor += snprintf(cursor,sizeof(logFormat) - (cursor - logFormat),"* Error * %s",format);*/
					n += fprintf(f, "[ERROR]");
					break;
				case LOG_WARNING:
					/*cursor += snprintf(cursor,sizeof(logFormat) - (cursor - logFormat),"* Warning * %s",format);*/
					n += fprintf(f, "[WARNING]");
					break;
				case LOG_NOTICE:
					/*cursor += snprintf(cursor,sizeof(logFormat) - (cursor - logFormat),"* Notice * %s",format); */
					n += fprintf(f, "[NOTICE]");
					break;
				case LOG_INFO:
					/*cursor += snprintf(cursor,sizeof(logFormat) - (cursor - logFormat),"* Info * %s",format);*/
					n += fprintf(f, "[INFO]");
					break;
				case LOG_DEBUG:
					/*cursor += snprintf(cursor,sizeof(logFormat) - (cursor - logFormat),"* Debug * %s",format);*/
					n += fprintf(f, "[DEBUG]");
					break;
				default:
					/*cursor += sprintf(cursor,"* <%d> * %s",priority,format);*/
					n += fprintf(f,"[<%d>]",priority);
				} /* switch(priority) */
			} /* (LogStat & LOG_LEVEL) */

			errno = saved_errno; /* restore errno for %m format */
			n += vfprintf(f, format, optional_arguments);

			/* Close the memory stream; this will finalize the data
		into a malloc'd buffer in BUF.  */
			fclose(f);
			/*
			 * begin of the critical section.
			 * Some of the function used below are thread cancellation points
			 * (according to Advanced Programming in the Unix Environment 2nd ed p411)
			 * so set the handler to avoid deadlocks and memory leaks
			 */
			cleanup_args cancelArgs;
			cancelArgs.buffer = buf;
			pthread_cleanup_push(cancel_handler, &cancelArgs);

			error = pthread_mutex_lock(&fileLock);
			if (likely(EXIT_SUCCESS == error)) {

				if (unlikely(LogStat & LOG_PERROR)) {
					/* add the format argument to the header */
					const size_t n = strlen(format);
					const size_t size = bufsize + n + 1;
					char *logFormatBuffer = (char *)malloc(size);
					if (logFormatBuffer) {
						pthread_cleanup_push(free,logFormatBuffer);
						strcpyNcat(logFormatBuffer,buf,format);
						vfprintf(stderr, logFormatBuffer, optional_arguments);
						pthread_cleanup_pop(1); /* pop the handler and free the allocated memory */
					} else {
						ERROR_MSG("failed to allocate %u bytes to print the msg on stderr",size);
					}
				} /* (LogStat & LOG_PERROR) */

				/* file management */
				if (unlikely(LOG_FILE_DURATION & LogStat)) {
					const time_t currentTime = time(NULL);
					const time_t elapsedTime = currentTime - startTime;
					if (unlikely(elapsedTime >= maxDuration)) {
						close(logFile);
						logFile = -1;
					}
				} /* (LOG_FILE_DURATION & LogStat) */

				if (unlikely(-1 == logFile)) {
					error = createFile();
					/* error already logged */
				}

				if (likely(EXIT_SUCCESS == error)) {
					const ssize_t written = write(logFile,buf,n);
					if (written > 0) {
						if (unlikely(written != n)) {
							ERROR_MSG("only %d byte(s) of %d has been written to %s",written,n,fullFileName);
							if (LogStat & LOG_CONS) {
								int fd = open("/dev/console", O_WRONLY | O_NOCTTY, 0);
								if (fd >= 0 ) {
									dprintf(fd,"logMsg");
									close(fd);
									fd = -1;
								}
							} /* (LogStat & LOG_CONS) */

							if (unlikely((LogStat & LOG_FILE_SYNC_ON_ERRORS_ONLY) && (LOG_PRI(priority) <= LOG_ERR))) {
								/* flush data if the log priority is "upper" or equal to error */
								error = fdatasync(logFile);
								if (error != 0) {
									error = errno;
									ERROR_MSG("fdatasync to %s error %d (%m)",fullFileName,error);
								}
								error = posix_fadvise(logFile, 0,0,POSIX_FADV_DONTNEED); /* tell the OS that log message bytes could be released from the file system cache */
								if (error != 0) {
								  ERROR_MSG("posix_fadvise to %s error %d (%m)",fullFileName,error);
								}
							} /* (unlikely((LogStat & LOG_FILE_SYNC_ON_ERRORS_ONLY) && (LOG_PRI(priority) <= LOG_ERR))) */
						} /* (unlikely(written != n)) */
						fileSize += written;

#ifdef FILESYSTEM_PAGE_CACHE_FLUSH_THRESHOLD
						static size_t currentPageCacheMaxSize = 0;
						currentPageCacheMaxSize += written;
						if (currentPageCacheMaxSize >= FILESYSTEM_PAGE_CACHE_FLUSH_THRESHOLD) {
							/* tell the OS that log message bytes could be released from the file system cache */
							if (likely(posix_fadvise(logFile, 0,0,POSIX_FADV_DONTNEED) == 0)) {
								currentPageCacheMaxSize = 0;
								DEBUG_MSG("used file system cache allowed to be flushed (size was %u)",currentPageCacheMaxSize);
							} else {
								NOTICE_MSG("posix_fadvise to %s error %d (%m), current page cache max size is %u",fullFileName,error,currentPageCacheMaxSize);
							}
						}
#endif /* FILESYSTEM_PAGE_CACHE_FLUSH_THRESHOLD */

						if ((LOG_FILE_ROTATE & LogStat) || (LOG_FILE_HISTO & LogStat)) {
							if (fileSize >= maxSize) {
								close(logFile);
								logFile = -1;
							}
						}
					} else if (0 == written) {
						WARNING_MSG("nothing has been written in %s", fullFileName);
					} else {
						error = errno;
						ERROR_MSG("write to %s error %d (%m)", fullFileName, error);
					}
				} /*(likely(EXIT_SUCCESS == error)) */

				/* End of critical section.  */
				/*pthread_cleanup_pop(0); implementation can't allow to set this instruction here */

				/* free allocated ressources */
				internalError = pthread_mutex_unlock(&fileLock);
				if (internalError != EXIT_SUCCESS) {
					ERROR_MSG("pthread_mutex_lock fileLock error %d (%s)", internalError, strerror(internalError));
					if (EXIT_SUCCESS == error) {
						error = internalError;
					}
				} /* (internalError != EXIT_SUCCESS) */
			} else {
				ERROR_MSG("pthread_mutex_lock fileLock error %d (%s)", error, strerror(error));
			}

			pthread_cleanup_pop(0);  /* moved to avoid a build error, use the preprocessor to understand why
			 * or have a look on man 3 pthread_cleanup_push:
			 * push & pop MUST BE in the SAME lexical nesting level !!!
			 */

			free(buf);
			cancelArgs.buffer = buf = NULL;
		} else { /* open_memstream(&buf, &bufsize) == NULL */
			ERROR_MSG("failed to get stream buffer");
			/* TMP! display the raw message to the console */
			vfprintf(stderr,format,optional_arguments);
			/* TODO: write the raw message to the file */
		}
	} /* ((LOG_MASK (LOG_PRI (priority)) & LogMask) != 0) */
	/* message has not to be displayed because of the current LogMask and its priority */

	/*return n;*/
}

void fileLogger(int priority, const char *format, ...) {
	va_list optional_arguments;
	va_start(optional_arguments, format);
	vfileLogger(priority, format,  optional_arguments);
	va_end(optional_arguments);
}

int setFileLoggerMaxSize_1_0(unsigned int maxSizeInBytes) {
	if (maxSizeInBytes != 0) {
		maxSize = maxSizeInBytes;
	}
	return maxSize;
}
/* function only available in the DSO to keep compatibility with the previous release */
#if defined(__pic__) || defined(__PIC__) || defined(PIC)
asm(".symver setFileLoggerMaxSize_1_0,setFileLoggerMaxSize@VERS_1.0");
#endif  /* defined(__pic__) || defined(__PIC__) || defined(PIC) */


int setFileLoggerMaxSize_1_1(const size_t maxSizeInBytes) {
	if (maxSizeInBytes != 0) {
		maxSize = maxSizeInBytes;
	}
	return maxSize;
}

/* new function available in the DSO and the static library */
#if defined(__pic__) || defined(__PIC__) || defined(PIC)
asm(".symver setFileLoggerMaxSize_1_1,setFileLoggerMaxSize@@VERS_1.1");
#else /* defined(__pic__) || defined(__PIC__) || defined(PIC) */
extern setFileLoggerMaxSize(size_t maxSizeInBytes) __attribute((alias("setFileLoggerMaxSize_1_1")));
#endif  /* defined(__pic__) || defined(__PIC__) || defined(PIC) */

int setFileLoggerMaxDuration(const time_t maxDurationInSeconds) {
	if (maxDurationInSeconds != 0) {
		maxDuration = maxDurationInSeconds;
	}
	return maxDuration;
}

int setFileLoggerDirectory(const char *logfileDirectory) {
	int error = EINVAL;

	if (logfileDirectory != NULL) {
		if (logfileDirectory[0] != '$') {
			error = checkDirectory(logfileDirectory);
		} else {
			char *envVar = getenv(logfileDirectory+1);
			if (envVar != NULL) {
				error = checkDirectory(logfileDirectory);
			} else {
				error = ENOENT;
				ERROR_MSG("environment variable %s is not found",logfileDirectory);
			}
		} /* !(logfileDirectory[0] != '$') */
		if (EXIT_SUCCESS == error) {
			strcpy(directory, logfileDirectory);
			DEBUG_VAR(directory, "%s");
			fileDirectory = directory;
		}
	} /* (logfileDirectory != NULL) */

	return error;
}

void closeLogFile(void) {
	if (logFile != -1) {
		close(logFile);
		logFile = -1;
	}
}

#include <dbgflags/ModuleVersionInfo.h>
MODULE_NAME(fileLogger);
MODULE_AUTHOR(Olivier Charloton);
MODULE_VERSION(1.0);
MODULE_FILE_VERSION(1.2);
MODULE_DESCRIPTION(logger to files with the same signature than syslog using POSIX API);
MODULE_COPYRIGHT(LGPL);
